adapters/richaudience/richaudience.go

```diff
@@ -0,0 +1,341 @@
+	if err := json.Unmarshal(responseData.Body, &bidResp); err != nil {
+
+		return nil, []error{&errortypes.BadServerResponse{
+			Message: err.Error(),
+		}}
+	}
+
+	if responseData.StatusCode != http.StatusOK {
+		err := &errortypes.BadServerResponse{
+			Message: fmt.Sprintf("Unexpected status code: %d. Run with request.debug = 1 for more info.", responseData.StatusCode),
+		}
+		return nil, []error{err}
+	}
+
+	var response openrtb2.BidResponse
+	if err := json.Unmarshal(responseData.Body, &response); err != nil {
+		return nil, []error{err}
+	}
+
+	bidResponse := adapters.NewBidderResponseWithBidsCapacity(len(request.Imp))
+	bidResponse.Currency = response.Cur
+	for _, seatBid := range response.SeatBid {
+		for _, bid := range seatBid.Bid {
+			b := &adapters.TypedBid{
+				Bid:     &bid,
+				BidType: "banner",
+			}
+			bidResponse.Bids = append(bidResponse.Bids, b)
+		}
+	}
+
+	return bidResponse, nil
+}
+
+func setHeaders(raiHeaders *http.Header) {
+	raiHeaders.Set("Content-Type", "application/json;charset=utf-8")
+	raiHeaders.Set("Accept", "application/json")
+	raiHeaders.Add("X-Openrtb-Version", "2.5")
+}
+
+func setImp(request *openrtb2.BidRequest, richaudienceRequest *richaudienceRequest) (resImps []openrtb2.Imp, err error) {
+	//Imp: Id, tagId, Secure, Bidfloor, Bidfloorcur, Banner, test
+	//Banner: Id, W, H, Format
+	//Format: W, H
+
+	for _, imp := range request.Imp {
+		var secure = int8(0)
+		raiExt := parseImpExt(&imp)
+
+		if raiExt != nil {
+			if raiExt.Pid != "" {
+				imp.TagID = raiExt.Pid
+			}
+
+			if raiExt.TestRa {
+				richaudienceRequest.Test = int8(1)
+			}
+
+			if raiExt.BidFloor <= 0 {
+				imp.BidFloor = 0.00001
+			} else {
+				imp.BidFloor = raiExt.BidFloor
+			}
+
+			if raiExt.BidFloorCur == "" {
+				imp.BidFloorCur = "USD"
+			} else {
+				imp.BidFloorCur = raiExt.BidFloorCur
+			}
+		}
+
+		if request.Site != nil && request.Site.Page != "" {
+			pageURL, error := url.Parse(request.Site.Page)
+			if error == nil && pageURL.Scheme == "https" {
+				secure = int8(1)
+			}
+		}
+		imp.Secure = &secure
+
+		if imp.Banner == nil {
+			err = &errortypes.BadInput{
+				Message: "Banner Object not found",
+			}
+			return
+		} else {
+			raiBanner := *imp.Banner
+			if raiBanner.W == nil && raiBanner.H == nil {
+				if len(raiBanner.Format) == 0 {
+					err = &errortypes.BadInput{
+						Message: "Format Object not found",
+					}
+				} else {
+					imp.Banner = &raiBanner
+				}
+			}
+		}
+
+		resImps = append(resImps, imp)
+
+	}
+	return
+}
+
+func setSite(request *openrtb2.BidRequest, richaudienceRequest *richaudienceRequest) {
+	if request.Site != nil && request.Site.Domain == "" {
+		if strings.Contains(request.Site.Page, "wwww.") && strings.Contains(request.Site.Page, "//") {
+			raiUrl := strings.Split(request.Site.Page, "//")
+			richaudienceRequest.Site.Domain = strings.Split(raiUrl[1], "/")[0]
+		}
+
+	} else {
+		richaudienceRequest.Site.Domain = request.Site.Domain
```

Comment:
This will fail  to NPE if request.Site is nil (in case request has App instead). I would move `if request.Site != nil` check to the top level and then check Domain and Page inside `if request.Site != nil`

Reply:
in next commit i have solved